File: src/meshcall/protocol.py
from __future__ import annotations

from typing import Annotated, Any, Literal, Self, TypeAlias

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    TypeAdapter,
    ValidationError,
    model_validator,
)

from meshcall.errors import ProtocolError
from meshcall.ir import BalancePolicy, StreamKind

PROTOCOL_VERSION = "meshcall/1"
Direction: TypeAlias = Literal["client", "server"]


class FrameModel(BaseModel):
    model_config = ConfigDict(extra="forbid")


class HelloFrame(FrameModel):
    kind: Literal["hello"] = "hello"
    protocol: str = PROTOCOL_VERSION
    role: Literal["client", "server"]
    peer_id: str
    instance_id: str | None = None


class HelloAckFrame(FrameModel):
    kind: Literal["hello.ack"] = "hello.ack"
    protocol: str = PROTOCOL_VERSION
    connection_id: str


class RegisteredMethod(FrameModel):
    name: str
    stream: StreamKind
    balance: BalancePolicy
    schema_hash: str = Field(pattern=r"^[0-9a-f]{64}$")


class RegisteredService(FrameModel):
    name: str
    methods: tuple[RegisteredMethod, ...]


class ServerRegisterFrame(FrameModel):
    kind: Literal["server.register"] = "server.register"
    instance_id: str
    services: tuple[RegisteredService, ...]


class MethodRegistrationResult(FrameModel):
    service: str
    method: str
    accepted: bool
    reason: Literal["schema_mismatch", "balance_disabled"] | None = None

    @model_validator(mode="after")
    def validate_outcome(self) -> Self:
        if self.accepted == (self.reason is not None):
            raise ValueError(
                "Accepted methods must omit a reason and rejected methods need one"
            )
        return self


class ServerRegisterAckFrame(FrameModel):
    kind: Literal["server.register.ack"] = "server.register.ack"
    instance_id: str
    methods: tuple[MethodRegistrationResult, ...]


class CallOpenFrame(FrameModel):
    kind: Literal["call.open"] = "call.open"
    call_id: str
    service: str
    method: str
    payload: Any
    deadline_unix_ms: int | None = None


class CallResultFrame(FrameModel):
    kind: Literal["call.result"] = "call.result"
    call_id: str
    payload: Any = None


class ErrorPayload(FrameModel):
    code: str
    message: str
    retryable: bool = False
    details: dict[str, Any] | None = None


class CallErrorFrame(FrameModel):
    kind: Literal["call.error"] = "call.error"
    call_id: str
    error: ErrorPayload


class CallCancelFrame(FrameModel):
    kind: Literal["call.cancel"] = "call.cancel"
    call_id: str
    reason: str = "cancelled"


class StreamItemFrame(FrameModel):
    kind: Literal["stream.item"] = "stream.item"
    call_id: str
    direction: Direction
    sequence: int = Field(ge=0)
    payload: Any


class StreamEndFrame(FrameModel):
    kind: Literal["stream.end"] = "stream.end"
    call_id: str
    direction: Direction


class StreamWindowFrame(FrameModel):
    kind: Literal["stream.window"] = "stream.window"
    call_id: str
    direction: Direction
    credit: int = Field(gt=0)


class PingFrame(FrameModel):
    kind: Literal["ping"] = "ping"
    nonce: str


class PongFrame(FrameModel):
    kind: Literal["pong"] = "pong"
    nonce: str


Frame: TypeAlias = Annotated[
    HelloFrame
    | HelloAckFrame
    | ServerRegisterFrame
    | ServerRegisterAckFrame
    | CallOpenFrame
    | CallResultFrame
    | CallErrorFrame
    | CallCancelFrame
    | StreamItemFrame
    | StreamEndFrame
    | StreamWindowFrame
    | PingFrame
    | PongFrame,
    Field(discriminator="kind"),
]

_FRAME_ADAPTER = TypeAdapter(Frame)


def encode_frame(frame: Frame) -> str:
    return _FRAME_ADAPTER.dump_json(frame, by_alias=True).decode()


def decode_frame(data: str | bytes) -> Frame:
    try:
        return _FRAME_ADAPTER.validate_json(data)
    except ValidationError as exc:
        raise ProtocolError(f"Invalid frame: {exc}") from exc

File: typescript/src/protocol.ts
import type { RawData } from "ws";

export const PROTOCOL_VERSION = "meshcall/1" as const;

export type StreamKind =
  | "unary"
  | "server_stream"
  | "client_stream"
  | "duplex";

export type BalanceKind =
  | "round_robin"
  | "least_inflight"
  | "random"
  | "sticky"
  | "disabled";

export interface BalancePolicy {
  kind: BalanceKind;
  key?: string | null;
}

export interface HelloFrame {
  kind: "hello";
  protocol: string;
  role: "client" | "server";
  peer_id: string;
  instance_id?: string | null;
}

export interface HelloAckFrame {
  kind: "hello.ack";
  protocol: string;
  connection_id: string;
}

export interface RegisteredMethod {
  name: string;
  stream: StreamKind;
  balance: BalancePolicy;
}

export interface RegisteredService {
  name: string;
  methods: Array<RegisteredMethod>;
}

export interface ServerRegisterFrame {
  kind: "server.register";
  instance_id: string;
  services: Array<RegisteredService>;
}

export interface ServerRegisterAckFrame {
  kind: "server.register.ack";
  instance_id: string;
}

export interface CallOpenFrame {
  kind: "call.open";
  call_id: string;
  service: string;
  method: string;
  payload: unknown;
  deadline_unix_ms?: number | null;
}

export interface CallResultFrame {
  kind: "call.result";
  call_id: string;
  payload?: unknown;
}

export interface ErrorPayload {
  code: string;
  message: string;
  retryable?: boolean;
  details?: Record<string, unknown> | null;
}

export interface CallErrorFrame {
  kind: "call.error";
  call_id: string;
  error: ErrorPayload;
}

export interface CallCancelFrame {
  kind: "call.cancel";
  call_id: string;
  reason?: string;
}

export interface StreamItemFrame {
  kind: "stream.item";
  call_id: string;
  direction: "client" | "server";
  sequence: number;
  payload: unknown;
}

export interface StreamEndFrame {
  kind: "stream.end";
  call_id: string;
  direction: "client" | "server";
}

export interface StreamWindowFrame {
  kind: "stream.window";
  call_id: string;
  direction: "client" | "server";
  credit: number;
}

export interface PingFrame {
  kind: "ping";
  nonce: string;
}

export interface PongFrame {
  kind: "pong";
  nonce: string;
}

export type Frame =
  | HelloFrame
  | HelloAckFrame
  | ServerRegisterFrame
  | ServerRegisterAckFrame
  | CallOpenFrame
  | CallResultFrame
  | CallErrorFrame
  | CallCancelFrame
  | StreamItemFrame
  | StreamEndFrame
  | StreamWindowFrame
  | PingFrame
  | PongFrame;

export function decodeFrame(data: RawData | string): Frame {
  const value: unknown = JSON.parse(typeof data === "string" ? data : data.toString());
  if (!isRecord(value) || typeof value.kind !== "string") {
    throw new Error("Invalid MeshCall frame");
  }
  return value as unknown as Frame;
}

export function encodeFrame(frame: Frame): string {
  return JSON.stringify(frame);
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
